jms/src/main/scala/akka/stream/alpakka/jms/Jms.scala

```diff
@@ -6,13 +6,29 @@ package akka.stream.alpakka.jms
 
 import java.time.Duration
 import javax.jms
-import javax.jms.ConnectionFactory
+import javax.jms.{ConnectionFactory, Message}
+
+case class AckEnvelope private[jms] (message: Message, private val jmsSession: JmsAckSession) {
+
+  @volatile var committed = false
+
+  def acknowledge(): Unit = if (!committed) jmsSession.ack(message)
+}
+
+case class TxEnvelope private[jms] (message: Message, private val jmsSession: JmsTxSession) {
+
+  @volatile var committed = false
+
+  def commit(): Unit = if (!committed) jmsSession.commit()
+
+  def rollback(): Unit = if (!committed) jmsSession.rollback()
+}
```

Comment:
Also, please let me know for any better ideas than `TxEnvelope` and `AckEnvelope` as type names for the message wrappers that provide safe ack and tx functionality from inside the stream.